src/ConstantInterval.cpp

```diff
@@ -0,0 +1,643 @@
+        result.max = std::max(result.max, v);
+    };
+
+    // Enumerate all possible values for the min and max and take the extreme values.
+    if (a.min_defined && b.min_defined && b.min != 0) {
+        consider_case(a.min, b.min);
+    }
+
+    if (a.min_defined && b.max_defined && b.max != 0) {
+        consider_case(a.min, b.max);
+    }
+
+    if (a.max_defined && b.max_defined && b.max != 0) {
+        consider_case(a.max, b.max);
+    }
+
+    if (a.max_defined && b.min_defined && b.min != 0) {
+        consider_case(a.max, b.min);
+    }
+
+    // Define an int64_t zero just to pacify std::min and std::max
+    constexpr int64_t zero = 0;
+
+    const bool b_positive = b > 0;
+    const bool b_negative = b < 0;
+    if ((b_positive && !b.max_defined) ||
+        (b_negative && !b.min_defined)) {
+        // Take limit as other -> +/- infinity
+        result.min = std::min(result.min, zero);
+        result.max = std::max(result.max, zero);
+    }
+
+    result.min_defined = ((a.min_defined && b_positive) ||
+                          (a.max_defined && b_negative));
+    result.max_defined = ((a.max_defined && b_positive) ||
+                          (a.min_defined && b_negative));
+
+    // That's as far as we can get knowing the sign of the
+    // denominator. For bounded numerators, we additionally know
+    // that div can't make anything larger in magnitude, so we can
+    // take the intersection with that.
+    if (a.is_bounded() && a.min != INT64_MIN) {
+        int64_t magnitude = std::max(a.max, -a.min);
+        if (result.min_defined) {
+            result.min = std::max(result.min, -magnitude);
+        } else {
+            result.min = -magnitude;
+        }
+        if (result.max_defined) {
+            result.max = std::min(result.max, magnitude);
+        } else {
+            result.max = magnitude;
+        }
+        result.min_defined = result.max_defined = true;
+    }
+
+    // Finally we can deduce the sign if the numerator and denominator are
+    // non-positive or non-negative.
+    bool a_non_negative = a >= 0;
+    bool b_non_negative = b >= 0;
+    bool a_non_positive = a <= 0;
+    bool b_non_positive = b <= 0;
+    if ((a_non_negative && b_non_negative) ||
+        (a_non_positive && b_non_positive)) {
+        if (result.min_defined) {
+            result.min = std::max(result.min, zero);
+        } else {
+            result.min_defined = true;
+            result.min = 0;
+        }
+    } else if ((a_non_negative && b_non_positive) ||
+               (a_non_positive && b_non_negative)) {
+        if (result.max_defined) {
+            result.max = std::min(result.max, zero);
+        } else {
+            result.max_defined = true;
+            result.max = 0;
+        }
+    }
+
+    // Normalize the values if it's undefined
+    if (!result.min_defined) {
+        result.min = 0;
+    }
+    if (!result.max_defined) {
+        result.max = 0;
+    }
+
```

Comment:
This function is complex enough that I think having a safety check like in other places would be good.

Reply:
i.e.
```
internal_assert(!result.is_bounded() || (result.min <= result.max)
```